cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i

```diff
@@ -0,0 +1,309 @@
+        bool isContainZero   = false;
+        for (size_t i = 0; i < k * n; ++i)
+        {
+            if (distances[i] < epsilon && distances[i] > -epsilon)
+            {
+                isContainZero = true;
+                break;
+            }
+        }
+
+        for (size_t i = 0; i < n; ++i)
+        {
+            for (size_t j = 0; j < nClasses; ++j)
+            {
+                classWeights[j] = 0;
+            }
+            for (size_t j = 0; j < k; ++j)
+            {
+                if (isContainZero)
+                {
+                    if (distances[i] < epsilon && distances[i] > -epsilon)
+                    {
+                        classWeights[trainLabel[indices[i * k + j]]] += 1;
+                    }
+                }
+                else
+                {
+                    classWeights[trainLabel[indices[i * k + j]]] += 1 / distances[i * k + j];
+                }
+            }
+            size_t maxWeightClass = 0;
+            FPType maxWeight      = 0;
+            for (size_t j = 0; j < nClasses; ++j)
+            {
+                if (classWeights[j] > maxWeight)
+                {
+                    maxWeight      = classWeights[j];
+                    maxWeightClass = j;
+                }
+            }
+            testLabel[i] = maxWeightClass;
+        }
+        return services::Status();
+    }
+
+    services::Status computeDistances(const NumericTable * trainTable, const NumericTable * testTable)
+    {
+        daal::SafeStatus s;
+
+        const size_t nDims  = trainTable->getNumberOfColumns();
+        const size_t nTrain = trainTable->getNumberOfRows();
+        const size_t nTest  = testTable->getNumberOfRows();
+
+        _distances = static_cast<FPType *>(services::internal::service_malloc<FPType, cpu>(nTrain * nTest * sizeof(FPType)));
+
+        daal::algorithms::internal::EuclideanDistances<FPType, cpu> euclDist(*testTable, *trainTable);
+        euclDist.init();
+
+        const size_t blockSize    = 128;
+        const size_t nOuterBlocks = nTest / blockSize + !!(nTest % blockSize);
+        const size_t nInnerBlocks = nTrain / blockSize + !!(nTrain % blockSize);
+
+        daal::threader_for(nOuterBlocks, nOuterBlocks, [&](size_t outerBlock) {
+            const size_t outerStart = outerBlock * blockSize;
+            const size_t outerEnd   = outerBlock + 1 == nOuterBlocks ? nTest : outerStart + blockSize;
+            const size_t outerSize  = outerEnd - outerStart;
+
+            daal::threader_for(nInnerBlocks, nInnerBlocks, [&](size_t innerBlock) {
+                const size_t innerStart = innerBlock * blockSize;
+                const size_t innerEnd   = innerBlock + 1 == nInnerBlocks ? nTrain : innerStart + blockSize;
+                const size_t innerSize  = innerEnd - innerStart;
+
+                daal::services::internal::TArray<FPType, cpu> tmpArr(outerSize * innerSize);
```

Comment:
Looks expensive to have it in the body of the loop. Can you have this allocation in per-thread basis?

Reply:
Removed in newer commits